In [3]:
# 1. Import Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import KFold, cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# 2. Load Dataset
df = pd.read_csv("Clean_Titanic_Data.csv")
df.head()

# 3. Define Features and Target
target = 'Pclass'  
X = df.drop(columns=[target])
y = df[target]

# 4. Model Training with K-Fold Cross-Validation (Before Tuning)
rf = RandomForestClassifier(random_state=42)
kf = KFold(n_splits=5, shuffle=True, random_state=42)

cv_scores = cross_val_score(rf, X, y, cv=kf, scoring='accuracy')
print("Pre-tuning Cross-Validation Accuracy Scores:", cv_scores)
print("Pre-tuning Mean Accuracy:", cv_scores.mean())

# 5. Hyperparameter Tuning using GridSearchCV
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'bootstrap': [True, False]
}

grid_search = GridSearchCV(
    estimator=rf,
    param_grid=param_grid,
    cv=kf,
    scoring='accuracy',
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X, y)
best_rf = grid_search.best_estimator_

print("Best Hyperparameters:", grid_search.best_params_)

# 6. Model Evaluation Post-Tuning
tuned_cv_scores = cross_val_score(best_rf, X, y, cv=kf, scoring='accuracy')
print("Post-tuning Cross-Validation Accuracy Scores:", tuned_cv_scores)
print("Post-tuning Mean Accuracy:", tuned_cv_scores.mean())

Pre-tuning Cross-Validation Accuracy Scores: [0.76536313 0.69662921 0.73033708 0.75280899 0.79213483]
Pre-tuning Mean Accuracy: 0.7474546481702342
Fitting 5 folds for each of 216 candidates, totalling 1080 fits
Best Hyperparameters: {'bootstrap': False, 'max_depth': 10, 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 100}
Post-tuning Cross-Validation Accuracy Scores: [0.75977654 0.71910112 0.73595506 0.74719101 0.79775281]
Post-tuning Mean Accuracy: 0.7519553072625699


**Reflection**

- **Pre-tuning Performance:**  
The Random Forest classifier achieved cross-validation accuracy scores of `[0.765, 0.697, 0.730, 0.753, 0.792]` across the 5 folds, with a mean accuracy of approximately **0.747**. This shows that the model was already performing reasonably well before tuning.

- **Hyperparameter Tuning:**  
  GridSearchCV tested 216 combinations of hyperparameters. The best parameters found were:  
  - `n_estimators`: 100  
  - `max_depth`: 10  
  - `min_samples_split`: 5  
  - `min_samples_leaf`: 1  
  - `bootstrap`: False  

- **Post-tuning Performance:**  
After tuning, cross-validation accuracy scores were `[0.760, 0.719, 0.736, 0.747, 0.798]` with a mean accuracy of approximately **0.752**. Hyperparameter tuning slightly improved overall predictive performance and made the model more consistent across folds.

- **Conclusion:**  
K-Fold cross-validation confirmed the model's reliability, and GridSearchCV successfully optimized Random Forest parameters, resulting in a more robust model for predicting passenger class (`Pclass`).